### ISSUE TITLE
uint32 overflow of sequence number – does seq has to be an integer?

### ISSUE THREAD

**Issue by regular** [flumedb/flumelog-offset#2]
While importing all of the npm registry into a flumelog-offset, I hit this:

```
TypeError('"value" argument is out of bounds');
"value" argument is out of bounds
at Buffer.writeUInt32BE (buffer.js:1120:5)
```

This of course is expected behaviour, the file cannot grow beyond 4Gb. I am aware that I could overcome this by using a log implementation that uses record numbers instead of the offset as the sequence number.

Anyway, this made me thinking, does the `sequence` have to be a number? Can it be a string or TypedArray? Shouldn't it be treated as opaque value by client code anyway? Could we wrap flumelog-offset into another log implementation called `flumelog-offset-multifile` or  `flumelog-offset-rolling`, that manages multiple offset files? A sequence could be a hex encoded number where the lower 32bit are the traditional file offset, and everything above is the file index, so we can spend a variable number of bytes on the file index (up to 4 obviously).

 ```
 offset = 0xff7f0001
4286513153
> fileIndex = 4
4
> b = Buffer(6)
<Buffer 00 00 00 00 00 00>
> b.writeUInt32LE(offset,0)
4
> b
<Buffer 01 00 7f ff 00 00>
> b.writeUInt16LE(fileIndex,4)
6
> b.toString('hex')
'01007fff0400'
```
This solution would also benefit users that are caught in crappy 32bit filesystems like FAT.

Any thoughts?

**Comment by dominictarr** [comment]
yup, most of those would work! but the simplest would just be to adapt the code to accept a byte width option (or encode/decode function) for the offset. just make it 64 bits, then you can do very very large files, although you can't actually do proper 64 bit numbers in javascript, you can do 53 bits, which is still very big. http://npm.im/uint53 (it still uses 64 bits, but js can only set 53 of them)

The great thing about using an _offset_ is that knowing the offset, you can get the record in _a single step_, O(1) where as other database schemes with a primary key, such as an incrementing sequence, like a BTree, need to traverse a binary structure, and so are O(log(n)) which is not as good as O(1)!!!

**Comment by regular** [comment]
Ok, seems to me it makes most sense to adapt https://github.com/flumedb/flumelog-offset/blob/master/frame/recoverable.js to take a filesizeCodec argument that can be queried for the byte width. This could then be either 8 byte / effectively 53 bits, [6 bytes / 48 bits](https://github.com/mafintosh/uint48be), or as it is now, 32 bits. With 48 bits, logs could grow to 256TB, which is probably fine for a while :)

**Comment by dominictarr** [comment]
could also use varints, it's just a little bit more complicated

**Comment by regular** [comment]
I implemented 48 and 53bit support and I want to test if the restore functionality still works. I am wondering, is there a reason why you break out of the restore tests [here](https://github.com/flumedb/flumelog-offset/blob/master/test/restore.js#L52) and [here](https://github.com/flumedb/flumelog-offset/blob/master/test/restore.js#L76)? They seem to work, if I remove the returns.

**Comment by regular** [comment]
Re: varints, yea, maybe in the future. I'm happy with the level of complexity, right now :)

**Comment by dominictarr** [comment]
oh those returns are not ment to be there. I just do that to bebug sometimes

**Comment by christianbundy** [comment]
Is this resolved by #6?

**Comment by dominictarr** [comment]
yes, also https://github.com/flumedb/flumelog-offset/issues/21
